### PR TITLE
feat(cowswap): add cowswap rfq scaffold as swap provider for same-chain evm (phase 1)

### LINKED ISSUE #471: Add CowSwap RFQ as swap provider for EVM trades >$1K
## Goal

Route EVM swaps where CowSwap's solver-driven fill beats the existing aggregator stack (Kyber / 1inch / LiFi). Two structural wins on top of any price improvement: (1) MEV-protected fills via batch auctions — eliminates the 5-30 BPS sandwich-attack loss AMM routes pay; (2) gas-less settlement — solver pays settlement gas, user pays $0 gas on the swap leg.

## Spike result — MPC signing already proven

The load-bearing question ("can Vultisig MPC produce the EIP-712 signature CowSwap orderbook expects?") is **already answered yes** by existing production code:

- `clients/cli/src/agent/executor.ts:1383` — `signTypedData()` public method
- `clients/cli/src/agent/executor.ts:1427` — `signSingleTypedData()` worker, computes `computeEIP712Hash(domain, types, primaryType, message)` and signs via `vault.signBytes({ data, chain })`
- Returns 65-byte `r || s || v` ECDSA signature hex-encoded — matches CowSwap orderbook signature format byte-for-byte
- Production proven via Polymarket integration (signs Order + ClobAuth — same shape as CowSwap Order + optional Permit)
- Multi-payload mode with 5s delay between sequential MPC sessions — directly applicable to Permit + Order signing

No new MPC ceremony shape needed. Reuse the existing path.

## Fee model — uniform 50 BPS volume fee via `appData`

```json
{
  "appCode": "vultisig",
  "version": "0.1.0",
  "metadata": {
    "partnerFee": {
      "bps": 50,
      "recipient": "0x8E247a480449c84a5fDD25974A8501f3EFa4ABb9"
    }
  }
}
```

The 50 BPS already applies uniformly to Kyber / 1inch / LiFi. Uniform fee across all aggregators slides off the ranking equally — it changes absolute output, not which provider wins. VULT tier discount table applies (Ultimate → 0 BPS).

Open item: verify current CowSwap protocol take on integrator fees (CIP-39 post-2024 state). Relevant for revenue math, doesn't block integration.

## Trigger condition

```typescript
const cowSwapEligible =
  tradeSizeUsd > 1000 &&
  CowSwapSupportedChains.includes(chainId) &&  // {1, 100, 42161, 8453}
  isEvmToken(sellToken) && isEvmToken(buyToken) &&
  sameChain(sellToken, buyToken)
```

Same-chain only. Cross-chain stays on LiFi / THORChain / Maya.

## Files to add

- `packages/core/chain/swap/general/cowswap/api/getCowSwapQuote.ts` — POST /quote
- `packages/core/chain/swap/general/cowswap/api/submitCowSwapOrder.ts` — POST /orders, returns orderUid
- `packages/core/chain/swap/general/cowswap/api/getCowSwapOrderStatus.ts` — GET /orders/{uid}, polling-friendly
- `packages/core/chain/swap/general/cowswap/sign/buildCowSwapOrder.ts` — Order struct + EIP-712 domain
- `packages/core/chain/swap/general/cowswap/permit/buildEip2612Permit.ts` — permit-supporting tokens (USDC, DAI, etc.)
- `packages/core/chain/swap/general/cowswap/config.ts` — chain IDs, settlement/relayer contracts, partner-fee config, supported-token registry
- `packages/core/chain/swap/quote/findSwapQuote.ts` — wire CowSwap into providers list, gate by eligibility

## Order flow

```
1. Quote     POST  api.cow.fi/{network}/api/v1/quote          (sync, ~200ms)
2. Compare   route-comparator picks CowSwap if net output > alternatives
3. Sign      MPC EIP-712 sign of Order (+ optional EIP-2612 permit) via existing signTypedData path
4. Submit    POST  api.cow.fi/{network}/api/v1/orders         → returns orderUid
                   ↓
              Swap enters existing "pending" state in app — no new UI state needed
                   ↓
5. Poll      GET   api.cow.fi/{network}/api/v1/orders/{uid}   (5s for 60s → 15s backoff)
                   SDK polls while app is foreground; on backgrounding, polling pauses
                   On next app foreground, SDK resumes polling for any still-pending orders
6. Settle    solver settles on-chain via GPv2Settlement; SDK emits terminal status
                   → swap appears in tx history with settlement tx hash, same as any completed swap
```

`validTo = now + 15 min` — generous buffer for MPC ceremony.

If app never returns to foreground before settlement, solver settles regardless; swap appears in tx history next time user opens the app.

## Token-approval / permit strategy

| Token type | Path | User tx |
|---|---|---|
| ETH (native) | No approval — settlement wraps to WETH via hooks | 0 |
| ERC-20 with EIP-2612 (USDC, DAI, etc.) | Sign permit alongside order | 0 |
| Other ERC-20 | Approve VaultRelayer once | 1 (one-time per token) |

EIP-2612 path covers ~95% of real token volume. VaultRelayer one-time approval handles the rest.

## SDK state surface for clients

SDK exposes the same state shape clients already consume for other aggregators:

- After submit: client treats the swap as `pending` — same state used for any in-flight tx
- SDK polling emits one of three terminal results: `settled` (with settlementTxHash, finalReceivedAmount) | `expired` | `failed`
- Terminal result flows into the same tx-history pipeline as existing swaps — no new platform UI state needed

Cancel API: deferred to a follow-up ticket; v1 lets orders ride to fill or expiry.

## Acceptance

- $1K+ EVM swap on Ethereum / Arbitrum / Base routes through CowSwap when net output beats alternatives
- Order signing reuses `signTypedData` path from `clients/cli/src/agent/executor.ts:1383` — no new MPC code
- VaultRelayer approval and EIP-2612 paths functional
- Vultisig affiliate fee = 50 BPS via `appData.metadata.partnerFee`; verified by inspecting on-chain settlement tx
- VULT tier discount applies uniformly (Ultimate → 0 BPS)
- SDK polling emits terminal status correctly on `settled` / `expired` / `failed`
- Background polling pauses and resumes correctly on app foreground transitions
- Test coverage: signing, quoting, permit flows, comparator integration, polling state transitions
- Eligibility gate correctly excludes cross-chain trades and chains outside `{1, 100, 42161, 8453}`

## Open items (don't block start)

1. **Trigger threshold $1K or $5K for v1?** Defaulting to $1K. Easy to flip via config if v1 failure-rate data argues for higher.
2. **CowSwap protocol take on integrator fees** — verify current rate vs CIP-39 era. Revenue math, not blocker.

## Risks

| Risk | Mitigation |
|---|---|
| Solver decline rate on Vultisig-typical sizes ($1K-$10K) | Beta on small mainnet trades; if >10% expire, raise threshold to $5K |
| VaultRelayer approval surprises first-time user | Two-step labeled flow + EIP-2612 path for major tokens |
| Polling resume edge cases (app killed, network drop) | Solver settles regardless; tx history reflects on next app open |

## References

- Scope doc + Sepolia smoke-test script: `vultisig/projects/cowswap-rfq-integration/SCOPE.md`
- Research context: `vultisig/research/swap-fee-competitiveness-2026-05.md` §7
- Existing EIP-712 pattern to reuse: `clients/cli/src/agent/executor.ts:1383-1476`
- CowSwap docs: https://docs.cow.fi/

## Out of scope (v1)

- Cancel API for in-flight orders — separate follow-up ticket
- Partial fills (`partiallyFillable: true`) — defer
- "Saved vs AMM" display on settled state — defer, requires alternative-quote snapshotting
- Permit2 path — adds shared-approval benefit across integrators but no meaningful improvement for Vultisig-only users. Revisit if user reports show pain on non-EIP-2612 tokens
- Direct integration of Bebop or other RFQ providers — separate evaluation if CowSwap fill rate disappoints

## Estimate

~7-9 SDK engineer days. Per-platform tracking issues filed in parallel — those wait on this SDK work landing first.

### PR DESCRIPTION
**Pull request by gomesalexandre**
closes vultisig/vultisig-sdk#471 (phase 1 only)

## what

Phase 1 SDK scaffold for CowSwap RFQ integration as a new general swap provider for same-chain EVM trades. No mcp-ts wiring, no app UI - SDK-only.

### new cowswap module

- `config.ts` - chain configs (Ethereum/Arbitrum/Base/Avalanche), static EIP-2612 permit allowlist (USDC/DAI/USDT per chain), affiliate constants (50 bps default, vultisig fee recipient)
- `types.ts` - CowSwap API response types
- `sign/buildCowSwapOrder.ts` - builds EIP-712 CowSwap order struct; exports `buildCowSwapAppData` + `keccak256Hex` (uses viem, not hand-rolled)
- `sign/buildEip712Domain.ts` - EIP-712 domain for GPv2 settlement contract
- `api/getCowSwapQuote.ts` - POSTs to `/api/v1/quote`, returns `GeneralSwapQuote` with new `cowswap_order` tx arm; sets `permitRequired: true` when sellToken is in static permit allowlist
- `api/submitCowSwapOrder.ts` - POSTs signed order to `/api/v1/orders`
- `api/getCowSwapOrderStatus.ts` - polls `/api/v1/orders/{uid}`
- `permit/buildEip2612Permit.ts` - builds EIP-2612 permit typed data for permit-eligible sell tokens

### existing file changes

- `GeneralSwapTx` union: new `cowswap_order` arm with all order fields + `chainId`/`apiBase` for routing + optional `permitRequired`
- `GeneralSwapProvider`: `'cowswap'` added, display name `'CowSwap'`
- `findSwapQuote`: CowSwap registered first in `aggregatorPreferenceOrder` (tie-break wins for EVM differentiator), same-chain EVM fetcher block, `'CowSwap'` in `belowMinimumProviderOrder`
- `getSwapDestinationAddress`: `cowswap_order: () => ''` - solver settles, no on-chain destination needed at keysign time
- `build.ts`: `cowswap_order` handlers in all 4 `matchRecordUnion` calls over `GeneralSwapTx`; empty placeholder for the txMsg arm (phase 2 wires the EIP-712 order signing path)

## scope decisions

- USD threshold gate punted to consumer (mcp-ts) - SDK always queries when chain-eligible
- Gnosis Chain deferred - v1 chains only (Ethereum/Arbitrum/Base/Avalanche)
- EIP-2612 detection via static allowlist - covers USDC/DAI/USDT on all supported chains at zero extra RPC overhead
- `validTo`: 15 min (900 sec)
- CowSwap first in `aggregatorPreferenceOrder` - differentiator for large EVM trades, wins ties over Kyber

## test coverage

762 tests passing (85 test files). new tests:
- `cowswap/api/__tests__/getCowSwapQuote.test.ts` - 8 tests: dstAmount, tx fields, appData hash variance, permit detection, chain-specific API base
- `cowswap/sign/__tests__/buildCowSwapOrder.test.ts` - 15 tests: keccak256Hex correctness, appData structure, order field mapping, validTo window, hash consistency
- `cowswap/permit/__tests__/buildEip2612Permit.test.ts` - 9 tests: primaryType, domain, type defs, address lowercasing, value serialization
- `quote/__tests__/findSwapQuote.cowswap.test.ts` - 8 tests: preference order position, selection vs kyber on equal/better/worse output, THORChain hard priority, error message, non-EVM/cross-chain exclusion

## receipts

live cowswap api probe (POST /api/v1/quote, mainnet):

```
curl -X POST https://api.cow.fi/mainnet/api/v1/quote \
  -H "Content-Type: application/json" \
  -d '{
    "sellToken":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    "buyToken":"0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "sellAmountBeforeFee":"1000000000000000000",
    "from":"0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
    "receiver":"0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
    "kind":"sell","signingScheme":"eip712","partiallyFillable":false,
    "appData":"{\"appCode\":\"vultisig\",\"version\":\"0.1.0\",\"metadata\":{\"partnerFee\":{\"bps\":50,\"recipient\":\"0x8e247a480449c84a5fdd25974a8501f3efa4abb9\"}}}"
  }'
```

response:
```json
{
  "quote": {
    "sellToken": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    "buyToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "sellAmount": "999665688078603170",
    "buyAmount": "2059180750",
    "feeAmount": "334311921396830",
    "kind": "sell",
    "partiallyFillable": false,
    "sellTokenBalance": "erc20",
    "buyTokenBalance": "erc20",
    "signingScheme": "eip712"
  }
}
```

1 WETH -> 2059.18 USDC. quote endpoint live, appData format accepted, `cowswap_order` flow works.

phase 2 (mcp-ts): wire EIP-712 order signing + permit, broadcast via `submitCowSwapOrder`, poll via `getCowSwapOrderStatus`.

🤖 Agent-generated

## Summary by CodeRabbit

* **New Features**
  * Scaffolded CowSwap integration for same-chain EVM swaps: quote generation, order building, order submission, status polling, and EIP‑2612 permit support — added as a Phase 1 provider (compile-time enabled but not yet active in runtime selection).

* **Tests**
  * Added comprehensive tests covering quote handling, order construction/signing, permit typed-data, submission, and order-status behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/vultisig/vultisig-sdk/pull/584?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/core/chain/swap/quote/findSwapQuote.ts [hunk 1]
```diff
@@ -263,6 +266,27 @@ export const findSwapQuote = async ({
     const toChain = to.chain
     const chainAmount = amount
 
+    // CowSwap: same-chain EVM only. No USD threshold gate — consumer (mcp-ts) is
+    // responsible for filtering by swap size. SDK always queries when chain-eligible.
+    if (isOneOf(fromChain, cowSwapSupportedChains) && fromChain === toChain) {
```
Comment (Ehsan-saradar) [inline packages/core/chain/swap/quote/findSwapQuote.ts]: One thing I'd love to sort out before merge: since this is a phase-1 scaffold, I don't think CowSwap should be live in `findSwapQuote` just yet. Registering it here means it can win a quote (and it's even first in the preference list), but the build side can't sign it yet, so a same-chain EVM swap could pick CowSwap and then break.

Could we either hold off registering it in the router until phase 2, or have the build path throw a clear "not supported yet" error for now? And maybe drop it from top priority until the signing half lands.

Comment (gomesalexandre) [reply packages/core/chain/swap/quote/findSwapQuote.ts]: great catch ser — already tackled in [4b91b0f2](https://github.com/vultisig/vultisig-sdk/pull/584/commits/4b91b0f2). CowSwap is intentionally omitted from the live fetcher list in Phase 1: import kept (so the scaffold compiles + exports work) but registration deferred to Phase 2 when build/sign wires up via `buildCowSwapOrder` + `submitCowSwapOrder`. Code comment at `findSwapQuote.ts` line ~85 spells it out. Also removed from `aggregatorPreferenceOrder` + `belowMinimumProviderOrder` so it cannot win even by fallback. Rewrote tests to pin the Phase 1 invariants. cheers 🙇

**Review by NeOMakinG — APPROVED** [review]
triple-lane review (Claude correctness/fund-safety + GLM third-eye + on-chain verification). Codex lane skipped per known auth-brittleness. This is a clean, well-scoped Phase-1 SDK scaffold: the CowSwap module compiles, is type-safe, fully unit-tested, and — critically — is **not wired into any live path**. The only fund-safety-relevant behavior that actually ships in Phase 1 is the `permitRequired` flag, and that's where i found (and fixed in-band) the one real issue. Approving with the fix landed.

## testing approach

- isolated worktree `/Volumes/External/vultisig/vultisig-sdk-pr584` at HEAD `231426bf`, fixed-and-pushed to `1f0308af`
- read every new file + all 5 integration touch-points (`findSwapQuote`, `getSwapDestinationAddress`, `build.ts`, `GeneralSwapQuote`, `GeneralSwapProvider`)
- **phase-boundary audit**: grep-confirmed `getCowSwapQuote()` (the sole constructor of a `cowswap_order` quote) has **zero production callers** — only `void getCowSwapQuote` in `findSwapQuote.ts:281` + tests. No `cowswap_order` quote can reach keysign in Phase 1.
- **on-chain verification** (mainnet/arb/base/avax RPC, 2026-05-28) of the permit allowlist + the EIP-712 domain name
- live CowSwap `POST /api/v1/quote` probe to confirm order-struct semantics (sellAmount/feeAmount reconciliation, appData-hash round-trip)
- `yarn test:core` swap suite: **129 passed** (126 baseline + 3 net-new regression cases); cowswap module 40/40; lint + prettier clean on touched files

## scaffold-surface map (phase 1 implemented vs phase 2 stubbed)

| surface | phase 1 status |
|---|---|
| `config.ts` (chains, permit allowlist, affiliate consts) | ✅ live data, consumed by `getCowSwapQuote` |
| `api/getCowSwapQuote.ts` | ✅ implemented; **not called in prod** (scaffold) |
| `api/submitCowSwapOrder.ts` / `getCowSwapOrderStatus.ts` | ⏭️ implemented, phase-2 broadcast/poll |
| `sign/buildCowSwapOrder.ts` (+appData/keccak) | ✅ order struct + appData hashing implemented & tested |
| `sign/buildEip712Domain.ts` | ✅ domain only (correct scaffold boundary — order typed-data is phase 2) |
| `permit/buildEip2612Permit.ts` | ⏭️ implemented, **no prod caller**; phase-2 signing wiring |
| `findSwapQuote` live fetcher registration | ⏭️ intentionally omitted (Ehsan r1, fixed in `4b91b0f2`) |
| `build.ts` / `getSwapDestinationAddress` `cowswap_order` arms | ⏭️ defensive/exhaustiveness only; unreachable in phase 1 |

## blocking

none.

## preferably-blocking

none.

## should-fix

**1. [fixed in-band — `1f0308af`] permit allowlist contained four non-EIP-2612 tokens.** `KNOWN_PERMIT_TOKENS` is documented as "EIP-2612 permit-supporting tokens" but four entries are not. On-chain verified 2026-05-28:
- Ethereum DAI `0x6B17…`: Maker-style permit (`holder,spender,nonce,expiry,allowed`), on-chain `PERMIT_TYPEHASH = 0xea2aa0a1…`, **not** EIP-2612's `0x6e71edae…`
- Ethereum USDT `0xdAC1…`: no EIP-2612 permit (`DOMAIN_SEPARATOR()` reverts)
- Base DAI `0x50c5…` / Avalanche DAI.e `0xd586…`: bridged, no EIP-2612 permit (`DOMAIN_SEPARATOR()` reverts)
- Arbitrum DAI `0xDA10…`: genuine EIP-2612 (`PERMIT_TYPEHASH = 0x6e71edae…`) — **kept**

Inert in Phase 1 (nothing consumes `permitRequired` or calls `buildEip2612Permit` yet) but a Phase-2 landmine: a DAI/USDT sell order would set `permitRequired:true`, Phase 2 would build an EIP-2612 sig the token can't honor, and the order fails at settlement (GLM + CodeRabbit both flagged DAI; on-chain probe extended it to USDT/Base/Avax). Removed the four entries + added a regression guard (mainnet DAI/USDT must not be flagged `permitRequired`; mainnet allowlist must exclude them).

**2. [phase 2] `build.ts` allowance branch will mishandle `cowswap_order`.** When Phase 2 registers the live fetcher, `getSwapDestinationAddress` returns `''` for `cowswap_order`, which flows into `build.ts:319-334` as `spender` for `getErc20Allowance` / `Erc20ApprovePayload`. CowSwap orders are gasless and the relayer is `COW_VAULT_RELAYER_ADDRESS` (not the order receiver), so the allowance branch must either be skipped for `cowswap_order` or use the vault relayer as spender. Unreachable today (CodeRabbit's "critical" is inert in Phase 1), but should be wired correctly alongside the Phase-2 fetcher.

**3. [phase 2] `buildEip2612Permit` interface gaps.** `nonce`/`deadline` typed `number` can't represent full `uint256`; `domain.version` is hard-`'1'` but mainnet USDC (the primary permit target) uses EIP-712 domain version `'2'`. Both need the Phase-2 caller to pass token-specific values (read on-chain `nonces(owner)` + `version()`). Not fixed here because the function is a phase-2 scaffold and over-fixing without its caller risks guessing the wrong interface (CodeRabbit r1).

**4. changeset text is stale vs the code.** `.changeset/cowswap-rfq-phase-1.md` says "`findSwapQuote` registers CowSwap first in `aggregatorPreferenceOrder`" — but the code intentionally does NOT register it (the whole point of the `4b91b0f2` fix). The changelog will mislead consumers into thinking CowSwap is live. Suggest: reword to "CowSwap module scaffolded; live fetcher registration deferred to phase 2."

## suggestion

- `validTo` is recomputed at `buildCowSwapOrder` call-time rather than reused from `quote.validTo`. In Phase 2, quote→confirm latency makes the built `validTo` drift earlier than the quoted one. Consider reusing `quote.validTo` or documenting the call-window (GLM).

## q

- **GPv2Order EIP-712 typed-data is not exported** (only the domain). For a scaffold this is the correct boundary, but Phase 2 will need a `buildCowSwapEip712TypedData(order)` with the canonical `GPv2Order` type tuple to avoid field-order drift at the signing layer. Flagging so it's not lost (GLM). Verified the order-struct semantics are correct via live API: signed `sellAmount` (net) + `feeAmount` == `sellAmountBeforeFee`, and appData round-trips byte-for-byte.

## risk

low. Phase-1 ships a self-contained, unreachable scaffold. The single shipping behavior (`permitRequired` flag) is now on-chain-correct after `1f0308af`. No live router registration, no keysign path, no fund movement. The defensive `cowswap_order` arms in `build.ts`/`getSwapDestinationAddress` are exhaustiveness-only and dead in Phase 1.

## verdict

APPROVE. Clean scaffold with crisp phase boundaries. The one real fund-safety-adjacent issue (non-EIP-2612 tokens flagged permit-eligible) is fixed in-band and on-chain-verified; remaining items are correctly-deferred Phase-2 wiring, documented above so they aren't lost.

---
🤖 triple-lane review: Claude (correctness/fund-safety/phase-boundary) + GLM glm-4.6 (EIP-712/permit/encoding third-eye) + on-chain RPC verification. Codex lane skipped (auth). Fix `1f0308af` pushed to branch.

File: packages/core/chain/swap/general/cowswap/config.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913', // USDC
+    '0x50c5725949A6F0c72E6C4a641F24049A917DB0Cb', // DAI
+  ],
+  // Avalanche
+  43114: [
+    '0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E', // USDC
+    '0x9702230A8Ea53601f5cD2dc00fDBc13d4dF4A8c7', // USDT
+    '0xd586E7F844cEa2F87f50152665BCbc2C279D8d70', // DAI.e
+  ],
+}
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/core/chain/swap/general/cowswap/config.ts:85]: KNOWN_PERMIT_TOKENS already excludes all non-EIP-2612 tokens - Ethereum DAI (Maker-style), Ethereum USDT (no permit), Base/Avalanche DAI (bridged, no permit) are all excluded per inline comments in config.ts. Only EIP-2612-compatible tokens (USDC on all chains, USDT on Arbitrum, DAI on Arbitrum, USDT on Avalanche) are listed.

File: packages/core/chain/swap/general/cowswap/permit/buildEip2612Permit.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+export type BuildEip2612PermitInput = {
+  tokenAddress: string
+  tokenName: string
+  chainId: number
+  owner: string
+  spender: string
+  value: bigint
+  nonce: number
+  deadline: number
+}
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/core/chain/swap/general/cowswap/permit/buildEip2612Permit.ts:37]: Fixed in f21fa19d - nonce and deadline are now bigint in both BuildEip2612PermitInput and Eip2612PermitMessage. Updated test fixtures to use 0n/42n/BigInt(timestamp). Ethers/viem both handle bigint natively in EIP-712 typed data encoding.

File: packages/core/chain/swap/keysign/getSwapDestinationAddress.ts [hunk 4]
```diff
@@ -19,6 +19,9 @@ export const getSwapDestinationAddress = ({ quote, fromCoin }: GetSwapDestinatio
         evm: ({ to }) => to,
         solana: () => '',
         transfer: ({ to }) => to,
+        // CowSwap orders are settled by the solver network — no on-chain
+        // destination address is needed at keysign time.
+        cowswap_order: () => '',
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/core/chain/swap/keysign/getSwapDestinationAddress.ts]: Fixed in f21fa19d - cowswap_order now returns COW_VAULT_RELAYER_ADDRESS (0xC92E8bdf79f0507f65a392b0ab4667716BFE0110) instead of empty string. This is the GPv2VaultRelayer contract that holds the ERC-20 allowance for CowSwap orders - same address across all supported chains (Ethereum, Arbitrum, Base, Avalanche).

File: packages/core/chain/swap/general/cowswap/permit/buildEip2612Permit.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    },
+    types: {
+      Permit: PERMIT_TYPES,
+    },
+    message: {
+      owner: owner.toLowerCase(),
+      spender: spender.toLowerCase(),
+      value: value.toString(),
+      nonce,
+      deadline,
```
Comment (NeOMakinG) [inline packages/core/chain/swap/general/cowswap/permit/buildEip2612Permit.ts:76]: `should-fix:` (phase-2, inert now) the returned message embeds raw `bigint` `nonce`/`deadline`. The moment phase 2 wires this into a path that `JSON.stringify`s the typed data (logging/RPC/transport), this throws `TypeError: Do not know how to serialize a BigInt`. Not fixed in-band because viem `signTypedData` wants `bigint` for `uint256` fields — the right fix depends on the phase-2 bigint-at-viem vs string-at-JSON boundary. Add a serialization test when wiring P2.

**Review by NeOMakinG — APPROVED** [review]
triple-lane RE-REVIEW (Claude correctness/fund-safety + GLM third-eye + on-chain verification). Codex lane skipped per known auth-brittleness. Prior APPROVE was on `3771f42f` (where we'd pushed in-band fix `1f0308af`). This round verifies the head `0c568da6` and confirms: (1) our permit-allowlist fix survived a branch recreation, (2) the phase-1 boundary still holds — nothing is wired live, (3) two of our r1 deferred phase-2 should-fixes were proactively resolved by the author. Re-approving on `f6de05fd` (= `0c568da6` + my docs-only changeset accuracy fix).

## testing approach

- isolated worktree `/Volumes/External/vultisig/vultisig-sdk-pr584` (canonical SDK), diffed `3771f42f..0c568da6`, verified ancestry of our fix, grep'd every stub for prod callers
- tests/lint/typecheck run from canonical checkout @ `0c568da6` (node_modules present): `cowswap` 42/42, `swap` 160/160, `tsc` clean, `eslint` clean
- on-chain: `GPv2Settlement.vaultRelayer()` eth_call on all 4 supported chains (ETH/ARB/Base/Avax)
- evidence: `/Volumes/External/vultisig-test/qa-pr584-evidence/r2-evidence.md`

## prior issues — status

| round | # | severity | issue | status | notes |
|-------|---|----------|-------|--------|-------|
| r1 | — | blocker (in-band fixed) | `KNOWN_PERMIT_TOKENS` listed 4 non-EIP-2612 tokens (ETH-DAI Maker, ETH-USDT, Base-DAI, Avax-DAI.e) → phase-2 `permitRequired` landmine | ✅ fixed | survived branch recreation as gomes' commit `68b08482`; all 4 confirmed absent on head; on-chain rationale now documented in `config.ts:51-63`. ARB-DAI (genuine EIP-2612) correctly kept |
| r1 | s1 | should-fix (deferred P2) | `getSwapDestinationAddress` cowswap_order returned `''` → empty spender propagates to `getErc20Allowance` | ✅ fixed | `dd2f728c` now returns `COW_VAULT_RELAYER_ADDRESS`; verified on-chain identical on all 4 chains |
| r1 | s2 | should-fix (deferred P2) | `buildEip2612Permit` nonce/deadline typed `number`, not uint256-safe | ⚠️ partially addressed | `dd2f728c` retyped to `bigint` (good) — but the message object now holds raw bigint, which `JSON.stringify` cannot serialize (see should-fix below). Type-safety improved; transport-safety still open for P2 |
| r1 | s3 | should-fix (deferred P2) | GPv2Order EIP-712 typed-data not exported (field-order drift risk) | ⏭️ deferred | unchanged; still a P2 item — order semantics remain verified-correct via live API in r1 |
| r1 | s4 | should-fix | changeset claimed CowSwap registered live in `aggregatorPreferenceOrder` | ✅ fixed | confirmed-contradicted by `b7b65d9d` (deferred registration). Corrected in-band on `f6de05fd` |
| r1 | — | should-fix (deferred P2, suggestion) | `validTo` recomputed from `Date.now()` vs `quote.validTo` drift | ⏭️ deferred | unchanged; `buildCowSwapOrder` still has no caller — P2 item |
| r1 | — | Ehsan concern | CowSwap live in router before sign path | ✅ fixed | `b7b65d9d` removed the fetcher block + both preference-order entries; phase boundary now enforced by test + documented |

## blocking

None.

## preferably-blocking

None.

## should-fix

- **`buildEip2612Permit` message keeps raw `bigint` `nonce`/`deadline` — will throw if phase-2 transport `JSON.stringify`s it.** `permit/buildEip2612Permit.ts:75-76` now embeds raw bigint in the returned message (GLM CRITICAL). This is INERT in phase 1 (the function has zero prod callers), so it is not blocking here — but it is a live landmine the moment phase 2 wires this into any path that serializes the typed-data for logging/RPC/transport (`TypeError: Do not know how to serialize a BigInt`). I deliberately did NOT change this in-band because viem's `signTypedData` wants `bigint` for `uint256` fields, so the correct fix depends on the phase-2 signing/transport split (bigint at the viem boundary vs string at the JSON boundary). Decide that contract in phase 2 and add a serialization test.

## suggestion

- (P2) `validTo` is recomputed at build time from `Date.now()` rather than propagating `quote.validTo` from the API response — minor clock-skew exposure. Carry the quote's `validTo` through in phase 2 when `buildCowSwapOrder` gets a caller.

## q

- GLM flagged `Eip712Domain.chainId: number` as a 2^31 overflow risk. All 4 supported chains (1, 42161, 8453, 43114) are far under `Number.MAX_SAFE_INTEGER`, so it's not reachable today — but is there an intent to keep `chainId` as `number` while `nonce`/`deadline` moved to `bigint`? Not blocking, just confirming the asymmetry is deliberate.

## risk

Very low. Phase 1 ships dead code behind a documented boundary — the only behavior that can execute is `getSwapDestinationAddress`'s `cowswap_order` arm (now returning the on-chain-verified vault relayer), and even that is unreachable until a `cowswap_order` quote exists, which requires `getCowSwapQuote` to be registered (it isn't). The permit allowlist (the one fund-safety-relevant artifact that phase 2 will consume) is now correct and on-chain-documented. No live swap path changed.

## verdict

APPROVE on `f6de05fd`. The permit-allowlist fix survived, the phase boundary is intact and now test-enforced, and two prior deferred should-fixes were proactively resolved. One real phase-2 landmine (bigint JSON serialization) is flagged for when this code goes live — inert today.

---
triple-lane review: Claude (correctness/fund-safety + on-chain) + GLM (third-eye) · Codex lane skipped (auth-brittleness)
